gufe/base.py

```diff
@@ -0,0 +1,381 @@
+        appropriate for the `GufeTokenizable` subclass.
+
+        """
+        sig = inspect.signature(self.__init__)
+
+        defaults = {
+            param.name: param.default for param in sig.parameters.values()
+            if param.default is not inspect.Parameter.empty
+        }
+
+        return defaults
+
+    @abc.abstractmethod
+    def _to_dict(self) -> Dict:
+        """This method should be overridden to provide the dict form of the
+        `GufeTokenizable` subclass.
+
+        `GufeTokenizable` instances shoulds *not* be used as keys in dicts
+        within this object; even though they are hashable, this makes
+        serialization into e.g. JSON difficult.
+
+        """
+        ...
+
+    @classmethod
+    @abc.abstractmethod
+    def _from_dict(cls, dct: Dict):
+        """This method should be overridden to receive the dict form of the
+        `GufeTokenizable` subclass and generate an instance from it.
+
+        """
+        ...
+
+    def to_dict(self, include_defaults=True) -> dict:
+        """Generate full dict representation, with all referenced
+        `GufeTokenizable` objects also given in full dict representations.
+
+        Parameters
+        ----------
+        include_defaults : bool
+            If `False`, strip keys from dict representation with values equal
+            to those in `defaults`.
+
+        See also
+        --------
+        :meth:`GufeTokenizable.to_shallow_dict`
+        :meth:`GufeTokenizable.to_keyed_dict`
+
+        """
+        dct = dict_encode_dependencies(self)
+
+        if not include_defaults:
+            for key, value in self.defaults.items():
+                if dct.get(key) == value:
+                    dct.pop(key)
+
+        return dct
+
+    @classmethod
+    def from_dict(cls, dct: Dict):
+        """Generate an instance from full dict representation.
+
+        Parameters
+        ----------
+        dct : Dict
+            A dictionary produced by `to_dict` to instantiate from.
+            If an identical instance already exists in memory, it will be
+            returned.  Otherwise, a new instance will be returned.
+
+        """
+        return dict_decode_dependencies(dct)
+
+    def to_keyed_dict(self) -> Dict:
+        return key_encode_dependencies(self)
+
+    @classmethod
+    def from_keyed_dict(cls, dct: Dict):
+        return key_decode_dependencies(dct)
+
+    def to_shallow_dict(self) -> Dict:
+        return to_dict(self)
+
+    @classmethod
+    def from_shallow_dict(cls, dct: Dict):
+        return from_dict(dct)
+
+
+class GufeKey(str):
+    def __repr__(self):
+        return f"<GufeKey('{str(self)}')>"
+
+    def to_dict(self):
+        return {':gufe-key:': str(self)}
+
+
+# TODO: may want to make this a weakref dict to avoid holding references here
+# why are we not using WeakValueDictionary?
+TOKENIZABLE_REGISTRY: Dict[str, weakref.ref[GufeTokenizable]] = {}
+# TOKENIZABLE_REGISTRY: Dict[str, GufeTokenizable] = WeakValueDictionary()
```

Comment:
As I mention in the comment; any reason not to use a `weakrefs.WeakValueDictionary`? Wouldn't that abstract out some of the explicit weakrefs stuff we're doing now?

Reply:
I didn't know about that, so can switch

Reply:
Switched to using `WeakValueDictionary`!